### PR TITLE
Iter5

### PR DESCRIPTION
**Pull request by Arcadian-Sky**

### CODE REVIEW

File: cmd/agent/main.go [hunk 1]
```diff
@@ -1,3 +1,176 @@
...
+)
+
+var (
+	serverAddress  = ""
+	pollInterval   = time.Second
+	reportInterval = time.Second
+)
+
+var (
+	pollCount float64
```
Comment (arttet) [inline cmd/agent/main.go:21]: CHANGE REQUEST: please don't use global variables

File: cmd/agent/main.go [hunk 2]
```diff
@@ -1,3 +1,176 @@
...
+	metrics   map[string]interface{}
+)
+
+// ADDRESS отвечает за адрес эндпоинта HTTP-сервера.
+// REPORT_INTERVAL позволяет переопределять reportInterval.
+// POLL_INTERVAL позволяет переопределять pollInterval.
+
+func main() {
+	end := flag.String("a", "localhost:8080", "endpoint")
```
Comment (arttet) [inline cmd/agent/main.go:30]: SUGGESTION: it would be better if you use `flags.go` for it.

File: cmd/agent/main.go [hunk 3]
```diff
@@ -1,3 +1,176 @@
...
+func updateMterics() {
+	for {
+		fmt.Println("updateMterics")
+		collectMetrics()
+		time.Sleep(pollInterval)
+	}
+}
+
+// Собираем метрики
+func collectMetrics() map[string]interface{} {
```
Comment (arttet) [inline cmd/agent/main.go:96]: CHANGE REQUEST: please split your application by layers

File: cmd/agent/main.go [hunk 4]
```diff
@@ -1,3 +1,176 @@
...
+var (
+	pollCount float64
+	metrics   map[string]interface{}
+)
+
+// ADDRESS отвечает за адрес эндпоинта HTTP-сервера.
+// REPORT_INTERVAL позволяет переопределять reportInterval.
+// POLL_INTERVAL позволяет переопределять pollInterval.
+
+func main() {
```
Comment (arttet) [inline cmd/agent/main.go:29]: CHANGE REQUEST: `main.go` should be like 

```
func main() {
	flag.Parse()
	collectAndSendMetrics()
}
```

File: internal/handler/handler.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
+package handler
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/Arcadian-Sky/musthave-metrics/internal/storage"
```
Comment (arttet) [inline internal/handler/handler.go:7]: CHANGE REQUEST: please format imports

https://stackoverflow.com/questions/38704769/golang-import-grouping-by-package

File: internal/handler/handler.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	metricType := chi.URLParam(r, "type")
+	metricName := chi.URLParam(r, "name")
+	metricValue := chi.URLParam(r, "value")
+
+	// fmt.Fprintf(w, "storage %s\n", metricType)
+	// fmt.Fprintf(w, "storage %s\n", metricName)
+	// fmt.Fprintf(w, "storage %s\n", metricValue)
+	fmt.Println("nonEmptyParts", metricType, metricName, metricValue, "|")
+	//Проверяем передачу типа
+	if metricType == "" {
```
Comment (arttet) [inline internal/handler/handler.go:36]: SUGGESTION: it should in the validate package

File: internal/handler/handler.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	metricType := chi.URLParam(r, "type")
+	metricName := chi.URLParam(r, "name")
+
+	// fmt.Fprintf(w, "storage %s\n", metricType)
+	// fmt.Fprintf(w, "storage %s\n", metricName)
+	fmt.Println("nonEmptyParts", metricType, metricName, "|")
+	//Проверяем передачу типа
+	if metricType == "" {
```
Comment (arttet) [inline internal/handler/handler.go:76]: CHANGE REQUEST: please avoid duplicate code

File: internal/server/server.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+		})
+	})
+	return r
+}
+
+func InitServer() {
+
+	r := InitRouter()
+
+	end := flag.String("a", ":8080", "endpoint address")
```
Comment (arttet) [inline internal/server/server.go:67]: CHANGE REQUEST: it would be better if it's in `cmd/server/flags.go`

**Review by arttet — CHANGES_REQUESTED** [review]
Please add 
- https://about.codecov.io/blog/getting-started-with-code-coverage-for-golang/
- https://github.com/golangci/golangci-lint

File: internal/handler/handler.go (hunk 5 above)
Comment (Arcadian-Sky) [reply internal/handler/handler.go:7]: this is grouping by vscode,it already makes grouping by types
Do you mean  separate local package from others like this?
```
import (
	"fmt"
	"net/http"

	"github.com/Arcadian-Sky/musthave-metrics/internal/storage"

	"github.com/go-chi/chi/v5"
)
```

Unfortunatelly, I don't understand how to make local repository name shorter, for example by alias
And is there any way to make an automatically separation 'goimports -local' doesn't work

File: .github/workflows/codecov.yml [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 2
+      - uses: actions/setup-go@v2
+        with:
+          go-version: '1.21'
+      - name: Run coverage
+        run: go test ./... -timeout 30s -race -coverprofile=coverage.out -covermode=atomic
+      - name: Upload coverage to Codecov
+        uses: codecov/codecov-action@v3
```
Comment (arttet) [inline .github/workflows/codecov.yml:18]:
```suggestion
        uses: codecov/codecov-action@v4
```

File: internal/server/handler/handler.go [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/go-chi/chi/v5"
+
+	"github.com/Arcadian-Sky/musthave-metrics/internal/server/storage"
+)
+
+func MetricsHandlerFunc(w http.ResponseWriter, r *http.Request) {
```
Comment (arttet) [inline internal/server/handler/handler.go:12]:
```suggestion
type Handler struct {
    s storage.Storage,
}

func (р *Handler) MetricsHandlerFunc(w http.ResponseWriter, r *http.Request) {
```

File: internal/server/handler/handler.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+	if storage.Storage == nil {
```
Comment (arttet) [inline internal/server/handler/handler.go:13]: CHANGE REQUEST: it should be constructed in the `main` package

File: internal/server/handler/handler.go [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+		storage.Storage = storage.NewMemStorage()
+	}
+	currentMetrics := storage.Storage.GetMetrics()
+	for name, value := range currentMetrics {
```
Comment (arttet) [inline internal/server/handler/handler.go:17]:
```suggestion
        // TODO: REMOVE ME
	for name, value := range currentMetrics {
```

File: internal/server/storage/storage.go [hunk 13]
```diff
@@ -0,0 +1,98 @@
+package storage
+
+import (
+	"fmt"
+	"strconv"
+)
+
+var Storage *MemStorage
```
Comment (arttet) [inline internal/server/storage/storage.go:8]: CHANGE REQUEST: please don't use global variables

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/agent/flags/flags.go [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+import (
+	"flag"
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+)
+
+var (
+	serverAddress  = ""
```
Comment (arttet) [inline internal/agent/flags/flags.go:12]: SUGGESTION: it shouldn't be a global variable

File: internal/server/handler/handler.go [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+	}
+}
+
+type Handler struct {
+	s storage.MetricsStorage
+}
+
+// NewHandler создает экземпляр Handler
+func NewHandler() *Handler {
+	return &Handler{}
```
Comment (arttet) [inline internal/server/handler/handler.go:34]:
```suggestion
	return &Handler{
	    s = myStorage
	}
```

File: internal/server/handler/handler.go [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+		Value: chi.URLParam(r, "value"),
+	}
+}
+
+type Handler struct {
+	s storage.MetricsStorage
+}
+
+// NewHandler создает экземпляр Handler
+func NewHandler() *Handler {
```
Comment (arttet) [inline internal/server/handler/handler.go:33]:
```suggestion
func NewHandler(myStorage Storage) *Handler {
```

File: internal/server/handler/handler.go [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+	return &Handler{}
+}
+
+func (h *Handler) InitStorage() {
```
Comment (arttet) [inline internal/server/handler/handler.go:37]: CHANGE REQUEST: please don't use the Singleton pattern

**Review by arttet — APPROVED** [review]

**Pull request merged**